### PR TITLE
Implement EnqueueExtensions interface in nodelabel

### PR DESCRIPTION
**Pull request by wzshiming**
#### What type of PR is this?

/kind feature
/sig scheduling

#### What this PR does / why we need it:

Implement EnqueueExtensions interface in nodelabel plugin.

#### Which issue(s) this PR fixes:

xref #94009

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/nodelabel/node_label.go [hunk 1]
```diff
@@ -135,3 +135,12 @@ func (pl *NodeLabel) Score(ctx context.Context, state *framework.CycleState, pod
 func (pl *NodeLabel) ScoreExtensions() framework.ScoreExtensions {
 	return nil
 }
+
+// EventsToRegister returns the possible events that may make a Pod
+// failed by this plugin schedulable.
+func (pl *NodeLabel) EventsToRegister() []framework.ClusterEvent {
+	return []framework.ClusterEvent{
+		{Resource: framework.Pod, ActionType: framework.Add},
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/nodelabel/node_label.go:143]: As this plugin's Filter implementation is only related to its plugin arguments - which labels to be present/absent. So a Pod event won't help in making this schedulable.

**Comment by Huang-Wei** [comment]
@wzshiming Building a benchmark test for `nodelabel` might be a bit challenging, as `nodelabel` plugin is not enabled by default. Let me know if you need help.

**Comment by wzshiming** [comment]
@Huang-Wei  
The `nodelabel` plugin can be activated in the benchmark now.  https://github.com/wzshiming/kubernetes/tree/fea/enq-nodelabel-tmp
2  bugs were found during debugging. please take a look if you have time. #99995 #100011
I will continue tomorrow.

**Comment by wzshiming** [comment]
1000 in new
```
+--------------------------+----------------------------------------+----------+--------+---------+
|          METRIC          |                 GROUP                  | QUANTILE |  UNIT  | MEDIAN  |
+--------------------------+----------------------------------------+----------+--------+---------+
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/1000Nodes | Average  | pods/s |   58.11 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/1000Nodes | Perc50   | pods/s |   62.50 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/1000Nodes | Perc90   | pods/s |   85.25 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/1000Nodes | Perc99   | pods/s |  149.50 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Average  | ms     |   33.46 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc50   | ms     |   11.35 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc90   | ms     |   78.31 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc99   | ms     |  143.88 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Average  | ms     | 4231.39 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc50   | ms     | 4228.42 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc90   | ms     | 8873.38 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc99   | ms     | 9556.69 |
+--------------------------+----------------------------------------+----------+--------+---------+
```

1000 in old

```
+--------------------------+----------------------------------------+----------+--------+---------+
|          METRIC          |                 GROUP                  | QUANTILE |  UNIT  | MEDIAN  |
+--------------------------+----------------------------------------+----------+--------+---------+
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/1000Nodes | Average  | pods/s |   56.32 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/1000Nodes | Perc50   | pods/s |   71.12 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/1000Nodes | Perc90   | pods/s |   90.33 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/1000Nodes | Perc99   | pods/s |  111.44 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Average  | ms     |   46.57 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc50   | ms     |   10.54 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc90   | ms     |   32.98 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc99   | ms     |  233.90 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Average  | ms     | 3902.42 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc50   | ms     | 3884.83 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc90   | ms     | 8430.19 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/1000Nodes | Perc99   | ms     | 9335.09 |
+--------------------------+----------------------------------------+----------+--------+---------+
```


5000 in new

```
+--------------------------+----------------------------------------+----------+--------+----------+
|          METRIC          |                 GROUP                  | QUANTILE |  UNIT  |  MEDIAN  |
+--------------------------+----------------------------------------+----------+--------+----------+
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/5000Nodes | Average  | pods/s |    59.09 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/5000Nodes | Perc50   | pods/s |    63.13 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/5000Nodes | Perc90   | pods/s |    77.25 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/5000Nodes | Perc99   | pods/s |   100.50 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Average  | ms     |    80.36 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc50   | ms     |    22.03 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc90   | ms     |    78.04 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc99   | ms     |   523.87 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Average  | ms     | 16546.82 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc50   | ms     | 16669.31 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc90   | ms     | 35087.33 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc99   | ms     | 38023.67 |
+--------------------------+----------------------------------------+----------+--------+----------+
```

5000 in old

```
+--------------------------+----------------------------------------+----------+--------+----------+
|          METRIC          |                 GROUP                  | QUANTILE |  UNIT  |  MEDIAN  |
+--------------------------+----------------------------------------+----------+--------+----------+
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/5000Nodes | Average  | pods/s |    58.77 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/5000Nodes | Perc50   | pods/s |    63.71 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/5000Nodes | Perc90   | pods/s |    78.29 |
| SchedulingThroughput     | SchedulingWithNodeLabelChurn/5000Nodes | Perc99   | pods/s |   105.57 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Average  | ms     |    96.47 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc50   | ms     |    20.69 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc90   | ms     |    95.66 |
| scheduler_e2e_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc99   | ms     |   627.75 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Average  | ms     | 16772.88 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc50   | ms     | 16647.89 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc90   | ms     | 34924.00 |
| scheduler_pod_scheduling | SchedulingWithNodeLabelChurn/5000Nodes | Perc99   | ms     | 37942.00 |
+--------------------------+----------------------------------------+----------+--------+----------+
```

**Comment by Huang-Wei** [comment]
https://github.com/kubernetes/kubernetes/pull/99937#issuecomment-794900346 doesn't look pretty good. Could you describe in detail how your test is built?

**Comment by wzshiming** [comment]
https://github.com/wzshiming/kubernetes/commit/d432239d2b63783c083e2ea513bee3e7eaed308d

Following
 - Enable plugin `nodelabel` with change code https://github.com/wzshiming/kubernetes/commit/d432239d2b63783c083e2ea513bee3e7eaed308d
 - Following https://gist.github.com/Huang-Wei/1d1d16ec0de0d23ce8ed2a2448a44c4d
 - Following https://github.com/Huang-Wei/k8s-sched-perf-stat

I designed a testcase like this

``` yaml
# config/node-with-nodelabel.yaml
apiVersion: v1
kind: Node
metadata:
  generateName: node-churn-scheduling-nodelabel-
  labels:
    foo: 'any'
    bar: 'any'
status:
  capacity:
    pods: "110"
    cpu: "4"
    memory: 32Gi
  conditions:
    - status: "True"
      type: Ready
  phase: Running
```

``` yaml
# config/pod-with-nodelabel.yaml
apiVersion: v1
kind: Pod
metadata:
  generateName: pod-churn-scheduling-nodelabel-
spec:
  schedulerName: nodelabel-scheduler
  containers:
  - image: k8s.gcr.io/pause:3.4.1
    name: pause
    resources:
      limits:
        cpu: 100m
        memory: 500Mi
      requests:
        cpu: 100m
        memory: 500Mi
```

``` yaml
# config/performance-config.yaml
- name: SchedulingWithNodeLabelChurn
  profiles: # Here is how I modified the test code and added the ability to enable plugins
  - schedulerName: nodelabel-scheduler
    plugins:
      filter:
        enabled:
        - name: NodeLabel
      score:
        enabled:
        - name: NodeLabel
          weight: 1
    pluginConfig:
    - name: NodeLabel
      args:
        presentLabels:
          - "foo"
        absentLabels: []
        presentLabelsPreference:
          - "bar"
        absentLabelsPreference: []
  workloadTemplate:
    - opcode: createNodes
      countParam: $initNodes
    - opcode: createPods
      countParam: $initPods
      podTemplatePath: config/pod-default.yaml
      skipWaitToCompletion: true
    - opcode: churn
      mode: create
      countParam: $initNodes
      templatePaths:
        - config/node-with-nodelabel.yaml
    - opcode: createPods
      countParam: $measurePods
      podTemplatePath: config/pod-with-nodelabel.yaml
      collectMetrics: true
  workloads:
    - name: 1000Nodes
      params:
        initNodes: 1000
        initPods: 2000
        measurePods: 1000
    - name: 5000Nodes
      params:
        initNodes: 5000
        initPods: 1000
        measurePods: 2000
```

**Comment by Huang-Wei** [comment]
The primary design principle of the test case is to ensure the unschedulabe Pods not gets moved upon the churn events. So I'd design the case like this:

- prepare a scheduler config with two profiles - one is `default-scheduler` (nothing get changes), the other is `node-label` where NodeLabel plugin and its pluginArgs need to be configured properly
- step1: `opcode: createNodes`. You don't need to apply labels to nodes created in this step.
- step2: `opcode: createPods`. The pods **need to be specified with `schedulerName: node-label` and with high-priority**, so that they are unschedulabel due to absence of nodelabels.
- step3: `opcode: churn`. Churn the cluster with irrelevant events, like PV/PVC/Service/StorageClass/etc.
- step4: `opcode: createPods`. Final step to measure Pods. Don't specify their `schedulerName` so they're not subject to profile `node-label` and hence schedulable.

I think this case can get a desired result.

**Comment by Huang-Wei** [comment]
I used to expect a benchmark test for each PR, as I'd like all PRs to be merged in 1.21. But we failed to achieve that, and now we have enough time to review all PRs in and let them soak for almost the whole 1.22 cycle, I'm not pursuing benchmark tests for each PR anymore.

So
/lgtm
/approve

But you can still build the benchmark test to verify the result difference, and now with #99638, it's easier to build a testcase with a customized scheduler config.

Thanks @wzshiming !

**Pull request merged**